# Welcome to Lab: Experimental Design 👨‍🔬⚗️🧪

Before this lab section, you completed the "Hello Survey" based on the questions you generated with your groups.  This week, you will do an analysis of the results and explore techniques to do great experimental design in Python!

A few tips to remember:

- Refer to your **lecture notebook** and the **pandas cheat sheet** to help you out with the code!
- If you find yourself stuck for more than a few minutes, ask a neighbor or course staff for help!  When you are giving help to your neighbor, explain the **idea and approach** to the problem without sharing the answer itself so they can have the same **<i>ah-hah</i>** moment!
- We are here to help you!  Don't feel embarrassed or shy to ask us for help!

Let's get started!

<hr style="color: #DD3403;">

In [ ]:
# Meet your CAs and TA if you haven't already!
# ...first name is enough, we'll know who they are! :)
ta_name = "Ramya"
ca1_name = "Vikram"
ca2_name = "Eliana"


# Say hello to each other!
# - Groups of 3 are ideal :)
# - However groups of 2 or 4 are fine too!

# NOTE: Team members from last week moved to another table (don't know why), couldn't get the names of the new team I randomly joined before they left lab.

partner1_name = ""
partner1_netid = ""
partner1_inYourGroupLastWeek = "yes"

partner2_name = ""
partner2_netid = ""
partner1_inYourGroupLastWeek = "yes"

partner3_name = ""
partner3_netid = ""
partner1_inYourGroupLastWeek = "yes"

<hr style="color: #DD3403;">

# Part 1: Hello Dataset Analysis

In lab last week, you created a series of questions that made up the "Hello Dataset".  Before lab, you completed this survey along with the entire course.  Today, you're going to do analysis on this dataset! :)


## Your Binary Question Analysis

You will be analyzing the **binary question** your group asked the class from `lab_pandas` -- record it below.  If multiple people from your group had different binary questions, decide which one to use and record that question below.

Do you primarily use Windows or Mac?

## Load the Hello Dataset

The "Hello Dataset" is available here:
```
https://waf.cs.illinois.edu/discovery/hello-fa22.csv
```

Use Python to load this dataset into a DataFrame called `df`:

In [3]:
import pandas as pd
df = pd.read_csv("https://waf.cs.illinois.edu/discovery/hello-fa22.csv")

## Find The Variable for Your Question

The command `df.columns.tolist()` will **list all of the columns (variables)** in a DataFrame.  Use `df.columns.tolist()` and read the output until you find the column name for your binary question.

- You may see the message: `Output exceeds the __size limit__. Open the full output data __in a text editor__`
- If you do, click the **in a text editor** link to open the whole list in a new tab. :)

In [4]:
df.columns.tolist()

['Name',
 'Lab',
 'Number 1 or Number 2?',
 'Number 3 or Number 5?',
 'What is your preferred way of saying one half?',
 '7UP or Sierra Mist?',
 'Android or iOS?',
 'Apple or Orange?',
 'Beef or Pork?',
 'Your birthday is...',
 'Do you prefer to wear black or white?',
 'Boba or Coffee?',
 'Cake or Pie?',
 'Canes or Chick-fil-a?',
 'Do you pour milk or cereal first?',
 'Do you prefer chocolate or vanilla?',
 'Coffee or Tea?',
 'Coke or Pepsi?',
 'Cookies or Ice Cream?',
 'Do you prefer ice cream in a cup or in a cone?',
 'Do you prefer dad or mom?',
 'Marvel or DC?',
 'Are you living in a dorm or an apartment?',
 'Your last digit in your NetID...',
 'Were you born in an even or odd month of the year?',
 'Would you rather have the superpower of flight or invisibility?',
 'UIUC Football or Basketball Team?',
 'French Fries or Tater Tots?',
 'Heads or Tails?',
 'Hot Weather or Cold Weather?',
 'Hot Dogs or Hamburgers?',
 'In-State or Out-of-State/International',
 'Introvert or Extrovert?',

## Record the Variable for Your Question

In the cell below, record your question's column name in the variable `analysis_column`.  You'll need that exact column name to reference data from that cell.

In [9]:
analysis_column = 'Do you primarily use Windows or Mac?'
analysis_column

'Do you primarily use Windows or Mac?'

### 🔬 Test Case Checkpoint 🔬

In [10]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "You do not have a DataFrame called `df`."
assert( analysis_column in df ), "Make certain your analysis_column is EXACTLY as it appears in the Hello Dataset."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Column Subset: View Only Your Question's Responses

To begin your question's analysis, you may want to **focus only on your question**.  In lecture, we have primarily focused on creating subsets of rows. However, you can also create a subset of columns!

To create a DataFrame with just a single column, you can use the syntax:
```
df[ ["column name"] ]
```
- Notice the **double set of square brackets**.  You need the set for `df`, and the second set for the column name.
- You can find this syntax under the "Subset Variables (Columns)" on your cheat sheet!

Us Python to display only the column you're analyzing.  This will make it easy to see what the two options were for your question:

In [12]:
df[['Do you primarily use Windows or Mac?']]

Do you primarily use Windows or Mac?
0                                    Mac
1                                    Mac
2                                    Mac
3                                    Mac
4                                Windows
..                                   ...
519                                  Mac
520                              Windows
521                              Windows
522                              Windows
523                              Windows

[524 rows x 1 columns]

## DataFrame: Group One

Create a DataFrame, `df1`, that contains everyone who picked the first option of your two answers.  *(Make sure to make this a subset of the original `df`, and not just your one column.)*

In [23]:
df1 = df[df['Do you primarily use Windows or Mac?'] == "Windows"]
df1

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
4            Reni  AYD                    2.0                    5.0   
8         Suvinay  AYP                    1.0                    5.0   
10        Maxwell  AYJ                    1.0                    5.0   
11         Ashley  AYN                    2.0                    5.0   
14      Tianxiang  AYR                    1.0                    5.0   
..            ...  ...                    ...                    ...   
518       Celesta  AYS                    1.0                    3.0   
520       Renault  AYR                    2.0                    5.0   
521  Shivam Patel  AYC                    1.0                    3.0   
522         Colin  AYI                    1.0                    5.0   
523          Evan  AYG                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
4                                            2-Jan                 7UP   
8                                            2-Jan                 7UP   
10                                           2-Jan                 7UP   
11                                             0.5         Sierra Mist   
14                                           2-Jan                 7UP   
..                                             ...                 ...   
518                                            0.5                 7UP   
520                                            0.5         Sierra Mist   
521                                            0.5                 7UP   
522                                          2-Jan         Sierra Mist   
523                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
4               iOS            Apple          Pork   On/Before June 30  ...   
8           Android           Orange           NaN     On/After July 1  ...   
10              iOS           Orange          Beef     On/After July 1  ...   
11              iOS            Apple          Beef   On/Before June 30  ...   
14          Android           Orange          Pork     On/After July 1  ...   
..              ...              ...           ...                 ...  ...   
518             iOS            Apple          Beef     On/After July 1  ...   
520         Android           Orange          Pork   On/Before June 30  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   
522             iOS           Orange          Beef   On/Before June 30  ...   
523         Android           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
4                                      Mocha   
8                                Frappuccino   
10                  Peach green tea lemonade   
11                         Caramel Macchiato   
14                                 Starbucks   
..                                       ...   
518                              Frappuccino   
520                                     moca   
521                          Don’t have one    
522                                      tea   
523                                      Tea   

    What is your favorite mobile app?  What is your favorite subject?  \
4                           Instagram                            Math   
8                             Netflix                       Astronomy   
10                            Spotify                            Math   
11                             Sudoku                            Math   
14                             Wechat                              CS   
..                                ...                             ...   
518                         instagram                         science   
520                           wechat                              cs    
521                           Youtube                  World History    
522

## DataFrame: Group Two

Create a new DataFrame, `df2`, that contains everyone who picked the other answer:

In [24]:
df2 = df[df['Do you primarily use Windows or Mac?'] == "Mac"]
df2

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0      CA Ana  AL1                    2.0                    5.0   
1      CA Ram  AYR                    2.0                    5.0   
2      CA Xin  AYO                    1.0                    5.0   
3       Lekha  AYH                    2.0                    2.0   
5      Linger  AYF                    2.0                    5.0   
..        ...  ...                    ...                    ...   
512  Veronica  AYO                    1.0                    3.0   
514      Bill  AYA                    1.0                    3.0   
516    Edward  AYQ                    1.0                    5.0   
517    Lauren  AYP                    1.0                    3.0   
519      Bill  AYC                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
5                                            2-Jan                 7UP   
..                                             ...                 ...   
512                                          2-Jan                 7UP   
514                                          2-Jan                 7UP   
516                                            0.5         Sierra Mist   
517                                          2-Jan                 7UP   
519                                          2-Jan         Sierra Mist   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
1               iOS            Apple           NaN     On/After July 1  ...   
2               iOS            Apple          Beef   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
5               iOS            Apple          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
512             iOS            Apple          Beef   On/Before June 30  ...   
514             iOS           Orange          Beef   On/Before June 30  ...   
516             iOS           Orange          Beef     On/After July 1  ...   
517             iOS            Apple          Beef   On/Before June 30  ...   
519             iOS            Apple          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
1                               Black coffee   
2                                  starbucks   
3                        pumpkin spice latte   
5                                 Starbucks    
..                                       ...   
512            Vanilla sweet cream cold brew   
514                                 Starbuck   
516                                    Philz   
517                                      tea   
519                                      NaN   

    What is your favorite mobile app? What is your favorite subject?  \
0                            Snapchat                           Math   
1                      Google chrome                     Mathematics   
2                           Starbucks                     Psychology   
3                             twitter              english (writing)   
5                          App Store                       Any STEM    
..                                ...                            ...   
512                        Pokemon Go                        History   
514                           Discord                           STAT   
516                       Xiaohongshu                        Finance   
517                         instagram                    

## Analysis

Discuss the following questions with your group and then write down your response!

**Q1: What percentage of people chose each of your two binary options?**

227/524 chose 'Windows'; roughly 43.3%

294/524 chose 'Mac'; roughly 56.1%

- NOTE: It appears that 3 people didn't answer the survey or didn't answer this particular question, as the total number of answers to our question totals to 521, not 524. Calculations can be re-done, but for the sake of simplicity the total 524 will be used. 

**Q2: Did this surprise you? Why or why not? Write at least one complete sentence.**

Yes, my assumption was that Windows had more flexibility and accessibility options when it came to coding, so I assume students who joined a coding-based course would be Windows users. While I myself am not a Mac user, I have looked into this and found that Mac actually offers a lot of resources nowadays for coders. 

**Q3: Do you think your question was the closest to a perfect 50/50 split among the whole class?**

There is a 12.8% discrepancy between both answers, so I wouldn't consider our question a perfect 50/50 split among the whole class, but it was relatively close.

<hr style="color: #DD3403;">

## Part 2: Creating Treatment and Control Groups

For the rest of this lab, we will consider the "Hello Dataset" to be a pre-survey of randomly selected participants for an upcoming experiment.  In lecture, you learned that a good experiment will split the participants into two groups: the "treatment" group and the "control" group.

Recall that it is necessary to have both groups:

- The "control" group is a comparison group that will receive a placebo and will act as a control to ensure that the treatment has an effect.
- The "treatment" group will receive a treatment and we will look for differences between the results of this treatment group and the control group.

## Part 2.1: Select a Treatment Group

Select half (50%) of the participants to be in the treatment group and place them into a DataFrame called `treatment`.
- Remember, you can use your **pandas cheat sheet** to help you out!

In [25]:
treatment = df.sample(frac = 0.5)
treatment

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
44         Kaylee   AYL                    2.0                    5.0   
403         Gavin   AYK                    1.0                    3.0   
355         Xinyi   AYL                    2.0                    5.0   
180          Josh   AYQ                    1.0                    3.0   
521  Shivam Patel   AYC                    1.0                    3.0   
..            ...   ...                    ...                    ...   
496       Arianna   AYN                    2.0                    5.0   
27           John   AYM                    2.0                    3.0   
426         Kibae   AYA                    1.0                    3.0   
332         Lanbu  BYOD                    1.0                    3.0   
199         Suraj   AL1                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
44                                           2-Jan                 7UP   
403                                            0.5         Sierra Mist   
355                                          2-Jan                 7UP   
180                                          2-Jan         Sierra Mist   
521                                            0.5                 7UP   
..                                             ...                 ...   
496                                          2-Jan                 7UP   
27                                           2-Jan         Sierra Mist   
426                                          2-Jan         Sierra Mist   
332                                          2-Jan                 7UP   
199                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
44              iOS           Orange          Pork     On/After July 1  ...   
403             iOS           Orange          Beef     On/After July 1  ...   
355             iOS           Orange          Beef   On/Before June 30  ...   
180             iOS           Orange          Beef   On/Before June 30  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
496             iOS           Orange          Beef   On/Before June 30  ...   
27              iOS            Apple          Pork   On/Before June 30  ...   
426             iOS            Apple          Beef     On/After July 1  ...   
332             iOS            Apple          Beef   On/Before June 30  ...   
199         Android            Apple           NaN   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
44                             Caramel latte   
403                                    Latte   
355                             Matcha Latte   
180                                    latte   
521                          Don’t have one    
..                                       ...   
496                                Cold brew   
27                             Hot chocolate   
426                                Starbucks   
332                                 hot coco   
199                              iced coffee   

    What is your favorite mobile app? What is your favorite subject?  \
44                            Discord              Political science   
403                          Snapchat                           Math   
355                          calendar                             CS   
180                           Youtube                           Math   
521                           Youtube                 World History    
..                                ...                            ...   
496                          DuoLingo                           Math   
27                          App Store                           Math   
426                          Youtuble                           Math   
3

## Part 2.2: Select a Control Group

Using the **exact same technique and syntax** as you did above, also select half (50%) of the participants to be in the control group.  Place them into a DataFrame called `control`. For now, it's okay if there are people who are in both groups.

In [26]:
control = df.sample(frac = 0.5)
control

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
22      Grace  AYD                    1.0                    3.0   
499      Gage  AYM                    2.0                    3.0   
515    Faayez  AYT                    2.0                    5.0   
221  Scarlett  AYJ                    2.0                    5.0   
91        Anh  AYC                    2.0                    5.0   
..        ...  ...                    ...                    ...   
457       Jay  AYA                    1.0                    3.0   
52    Zhifeng  AYI                    2.0                    3.0   
200    james   AYF                    2.0                    5.0   
115    Jeremy  AYO                    1.0                    3.0   
387    Howard  AYK                    2.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
22                                           2-Jan                 7UP   
499                                            0.5                 7UP   
515                                          2-Jan                 7UP   
221                                            NaN                 NaN   
91                                           2-Jan                 7UP   
..                                             ...                 ...   
457                                          2-Jan                 7UP   
52                                           2-Jan                 7UP   
200                                          2-Jan                 7UP   
115                                          2-Jan                 7UP   
387                                            0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
22              iOS            Apple          Beef   On/Before June 30  ...   
499             iOS           Orange          Beef   On/Before June 30  ...   
515         Android            Apple          Beef   On/Before June 30  ...   
221             iOS           Orange          Beef     On/After July 1  ...   
91          Android            Apple          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
457             iOS            Apple          Beef     On/After July 1  ...   
52              iOS           Orange          Beef   On/Before June 30  ...   
200             iOS            Apple          Beef   On/Before June 30  ...   
115             iOS            Apple          Beef   On/Before June 30  ...   
387             iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
22                   Iced Caramel Macchiato    
499                            Dunkin Donuts   
515                               Chai latte   
221                            hot chocolate   
91                                 Black tea   
..                                       ...   
457                                   frappe   
52                                    Coffee   
200                                 starbuck   
115                              Apple juice   
387                         Iced Cappuccino    

    What is your favorite mobile app? What is your favorite subject?  \
22                           Snapchat                           Math   
499                          Snapchat                           Math   
515                           Spotify                      Economics   
221                               NaN                            NaN   
91                            Spotify               Computer Science   
..                                ...                            ...   
457                              mint                           Math   
52                           Coursera                           Math   
200                            chrome                             cs   
115                           Youtube               Compu

### 🔬 Test Case Checkpoint 🔬

In [27]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "You do not have a DataFrame called `df`."
assert( 'treatment' in vars() ), "You do not have a DataFrame called `treatment`."
assert( 'control' in vars() ), "You do not have a DataFrame called `control`."
assert( len(treatment) == len(control) ), "The `treatment` and `control` groups must be the same size."
assert( abs((len(df) / 2) - len(treatment)) <= 1 ), "The `treatment` group size must be 50% of the `df`"
assert( abs((len(df) / 2) - len(control)) <= 1 ), "The `control` group size must be 50% of the `df`"


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Analysis: Overlapping Selection?

The following code will detect rows that are **both in your treatment and control** groups.

Run the following cell to see everyone who is in both groups:

In [28]:
df[ df.index.isin(treatment.index) & df.index.isin(control.index) ]

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
1     CA Ram  AYR                    2.0                    5.0   
2     CA Xin  AYO                    1.0                    5.0   
3      Lekha  AYH                    2.0                    2.0   
10   Maxwell  AYJ                    1.0                    5.0   
12     Aimee  AYM                    1.0                    3.0   
..       ...  ...                    ...                    ...   
507    Laiba  AYO                    1.0                    5.0   
515   Faayez  AYT                    2.0                    5.0   
518  Celesta  AYS                    1.0                    3.0   
519     Bill  AYC                    1.0                    5.0   
523     Evan  AYG                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
1                                            2-Jan                 7UP   
2                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
10                                           2-Jan                 7UP   
12                                           2-Jan                 7UP   
..                                             ...                 ...   
507                                            0.5                 7UP   
515                                          2-Jan                 7UP   
518                                            0.5                 7UP   
519                                          2-Jan         Sierra Mist   
523                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
1               iOS            Apple           NaN     On/After July 1  ...   
2               iOS            Apple          Beef   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
10              iOS           Orange          Beef     On/After July 1  ...   
12              iOS            Apple          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
507             iOS            Apple          Beef     On/After July 1  ...   
515         Android            Apple          Beef   On/Before June 30  ...   
518             iOS            Apple          Beef     On/After July 1  ...   
519             iOS            Apple          Beef     On/After July 1  ...   
523         Android           Orange          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
1                               Black coffee   
2                                  starbucks   
3                        pumpkin spice latte   
10                  Peach green tea lemonade   
12                 Cold brew (Pumpkin cream)   
..                                       ...   
507                                Starbucks   
515                               Chai latte   
518                              Frappuccino   
519                                      NaN   
523                                      Tea   

    What is your favorite mobile app?  What is your favorite subject?  \
1                      Google chrome                      Mathematics   
2                           Starbucks                      Psychology   
3                             twitter               english (writing)   
10                            Spotify                            Math   
12                            twitter                 Social science    
..                                ...                             ...   
507                          Snapchat                              cs   
515                           Spotify                       Economics   
518                         instagram                         science   
519                               NaN                     

**Q: Why would your code result in some people being included in both groups? Write at least one complete sentence explaining why you think this might occur.**

The code that was used randomly picked out 50% of participants in the survey, that is all it was functioned to do. No restrictions were made alongside this code, therefore the code ran didn't ensure that people in either group were unique to the treatment and control group. 

## Part 2.3:

Read the following guide on **using `.index.isin` to select all rows from a DataFrame not in another DataFrame**: https://discovery.cs.illinois.edu/guides/DataFrame-Row-Selection/dataframe-isin-selection/

Finally, create a new `control` group that contains all of the participants from `df` that are not in `treatment`:

In [30]:
control = df[~df.index.isin(treatment.index)]
control

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0      CA Ana   AL1                    2.0                    5.0   
4        Reni   AYD                    2.0                    5.0   
6    Esbeida    AYA                    2.0                    5.0   
7     Jocelyn   AYS                    1.0                    5.0   
8     Suvinay   AYP                    1.0                    5.0   
..        ...   ...                    ...                    ...   
513       Jen   AYC                    2.0                    5.0   
514      Bill   AYA                    1.0                    3.0   
517    Lauren   AYP                    1.0                    3.0   
520   Renault   AYR                    2.0                    5.0   
522     Colin   AYI                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
4                                            2-Jan                 7UP   
6                                            2-Jan                 7UP   
7                                            2-Jan         Sierra Mist   
8                                            2-Jan                 7UP   
..                                             ...                 ...   
513                                          2-Jan                 7UP   
514                                          2-Jan                 7UP   
517                                          2-Jan                 7UP   
520                                            0.5         Sierra Mist   
522                                          2-Jan         Sierra Mist   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
4               iOS            Apple          Pork   On/Before June 30  ...   
6               iOS           Orange          Beef     On/After July 1  ...   
7               iOS           Orange          Beef     On/After July 1  ...   
8           Android           Orange           NaN     On/After July 1  ...   
..              ...              ...           ...                 ...  ...   
513         Android            Apple          Pork     On/After July 1  ...   
514             iOS           Orange          Beef   On/Before June 30  ...   
517             iOS            Apple          Beef   On/Before June 30  ...   
520         Android           Orange          Pork   On/Before June 30  ...   
522             iOS           Orange          Beef   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
4                                      Mocha   
6                       Iced vanilla coffee    
7                Caramel crunch Frappuccino    
8                                Frappuccino   
..                                       ...   
513                                     None   
514                                 Starbuck   
517                                      tea   
520                                     moca   
522                                      tea   

    What is your favorite mobile app? What is your favorite subject?  \
0                            Snapchat                           Math   
4                           Instagram                           Math   
6                           Snapchat                            Math   
7                             Spotify                        History   
8                             Netflix                      Astronomy   
..                                ...                            ...   
513                           Twitter                           Art    
514                           Discord                           STAT   
517                         instagram                        english   
520                           wechat         

### 🔬 Test Case Checkpoint 🔬

In [31]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df' in vars() ), "You do not have a DataFrame called `df`."
assert( 'treatment' in vars() ), "You do not have a DataFrame called `treatment`."
assert( 'control' in vars() ), "You do not have a DataFrame called `control`."
assert( abs((len(df) / 2) - len(treatment)) <= 1 ), "The `treatment` group size must be 50% of the `df`"
assert( abs((len(df) / 2) - len(control)) <= 1 ), "The `control` group size must be 50% of the `df`"
assert( len(df[ df.index.isin(treatment.index) & df.index.isin(control.index) ]) == 0), "The `treatment` and `control` groups have some overlap!  There must be no overlap."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 2.4: Finding Yourself!

Use the following cells to write Python code to find yourself in either the treatment group or control group.
- **Hint**: How do you select a subset of rows?  What conditional will identify you??
- If you did not fill out the "Hello Survey" then you are not in this dataset. :(
- If you are not in this survey, find either "Prof. Wade" or "Prof. Karle" instead.  *(Those are the names we used in the `Name` column.)*

In [38]:
# Use this cell to find yourself in either `treatment` or `control` groups:
myself = control[(control['Name'] == 'John') & (control['Lab'] == 'AYS')]
myself

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
475  John  AYS                    1.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
475                                          2-Jan                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
475             iOS           Orange          Beef   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
475                                Cold brew   

    What is your favorite mobile app? What is your favorite subject?  \
475                        Pokemon Go                      Astronomy   

    Who is your favorite singer? What is your favorite season?  \
475                Billie EIlish                          Fall   

    What is your Zodiac sign? What is your favorite video game?  \
475                 Capricorn             The Last of Us Series   

    What is your favorite non-video game? What is your favorite food?  \
475                                TikTok              Boneless Wings   

    What's your favorite movie?  
475                Harry Potter  

[1 rows x 120 columns]

## Analysis: Finding Yourself!

**Q: Which group where you in -- treatment or control?**


Control group

<hr style="color: #DD3403;">

# Part 3: Using Blocking to Create Treatment and Control Groups that are Balanced

In your pre-experiment research, you have reason to believe that how the participants answer **your group's question** might influence the response of this experiment!  Oh no! Even though this is a large sample of participants, we want to block to be absolutely sure that the treatment and control groups are balanced with regard to your group's question!

To control for this, we will using **blocking**!  Remember from lecture that blocking has us creating subgroups for each response, selecting equal-sized treatment and control groups from each block, and then combining all the treatment groups together and all the control groups together to get the final groups. This ensures that your groups are balanced with regard to the variable you blocked on!

You have already done some of this work:
- Earlier in this notebook, you created `df1` that includes everyone who answered your group's question one way.
- Earlier in this notebook, you created `df2` that includes everyone who answered your group's question the other way.

## Part 3.1:

Create a `df1_treatment` group that contains a random selection of 50% of the rows from `df1`.

In [39]:
df1_treatment = df1.sample(frac = 0.5)
df1_treatment

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
482    Tucker   AYB                    1.0                    5.0   
102  Shengwen   AYN                    2.0                    5.0   
420    Amanda   AYB                    2.0                    5.0   
207     Jimmy   AYC                    1.0                    5.0   
230   Pickles   AYC                    2.0                    5.0   
..        ...   ...                    ...                    ...   
107      John   AYK                    1.0                    5.0   
112        ek   AYM                    1.0                    3.0   
344  Anushree   AYH                    2.0                    5.0   
399      Feng  Stat                    1.0                    3.0   
439     Mason  AYQ                     1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
482                                            0.5                 7UP   
102                                          2-Jan                 7UP   
420                                          2-Jan         Sierra Mist   
207                                          2-Jan                 NaN   
230                                            0.5                 7UP   
..                                             ...                 ...   
107                                          2-Jan                 7UP   
112                                          2-Jan                 7UP   
344                                            0.5         Sierra Mist   
399                                          2-Jan                 7UP   
439                                            0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
482             iOS           Orange          Beef     On/After July 1  ...   
102             iOS            Apple          Pork   On/Before June 30  ...   
420             iOS           Orange          Beef   On/Before June 30  ...   
207             iOS           Orange          Beef     On/After July 1  ...   
230             iOS            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
107         Android           Orange          Beef   On/Before June 30  ...   
112         Android           Orange          Beef   On/Before June 30  ...   
344             iOS           Orange          Pork   On/Before June 30  ...   
399             iOS            Apple          Beef   On/Before June 30  ...   
439         Android            Apple          Beef   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
482                                  frappe    
102                              Cappuccino    
420                                   matcha   
207                                      NaN   
230                                      Tea   
..                                       ...   
107                                    Mocha   
112                                No coffee   
344                               cold brew    
399                                       No   
439                                    Water   

    What is your favorite mobile app? What is your favorite subject?  \
482                            TikTok                           Math   
102                            Canvas               Computer science   
420                         smash hit                       business   
207                            WeChat                        Physics   
230                      Illinois app                        English   
..                                ...                            ...   
107                         Instagram                           Math   
112                               NaN                          Maths   
344                            BeReal                           Math   
399                            Tiktok        

## Part 3.2:

Create a `df1_control` group that contains the other half of the rows from `df1`. *(You can refer back to Part 2 to refresh on how to do this, make sure you use `df1` and `df1_control` as your DataFrames.)*

In [40]:
df1_control = df1[~df1.index.isin(df1_treatment.index)]
df1_control

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
14      Tianxiang  AYR                    1.0                    5.0   
16        Claudia  AL1                    2.0                    3.0   
17           Wing  AYB                    1.0                    5.0   
24          Brian  AYF                    1.0                    3.0   
25      Shreshika  AYB                    2.0                    3.0   
..            ...  ...                    ...                    ...   
496       Arianna  AYN                    2.0                    5.0   
499          Gage  AYM                    2.0                    3.0   
515        Faayez  AYT                    2.0                    5.0   
520       Renault  AYR                    2.0                    5.0   
521  Shivam Patel  AYC                    1.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
14                                           2-Jan                 7UP   
16                                           2-Jan         Sierra Mist   
17                                           2-Jan         Sierra Mist   
24                                             0.5                 7UP   
25                                           2-Jan                 7UP   
..                                             ...                 ...   
496                                          2-Jan                 7UP   
499                                            0.5                 7UP   
515                                          2-Jan                 7UP   
520                                            0.5         Sierra Mist   
521                                            0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
14          Android           Orange          Pork     On/After July 1  ...   
16              iOS            Apple          Beef     On/After July 1  ...   
17              iOS            Apple          Beef   On/Before June 30  ...   
24              iOS            Apple          Beef   On/Before June 30  ...   
25              iOS           Orange           NaN   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
496             iOS           Orange          Beef   On/Before June 30  ...   
499             iOS           Orange          Beef   On/Before June 30  ...   
515         Android            Apple          Beef   On/Before June 30  ...   
520         Android           Orange          Pork   On/Before June 30  ...   
521             iOS            Apple           NaN   On/Before June 30  ...   

              What is your favorite coffee shop drink?  \
14                                           Starbucks   
16   Ice caramel macchiato or salted caramel cream ...   
17                                  Bridgeport Coffees   
24                                          Cappuccino   
25                                                 NaN   
..                                                 ...   
496                                          Cold brew   
499                                      Dunkin Donuts   
515                                         Chai latte   
520                                               moca   
521                                    Don’t have one    

    What is your favorite mobile app? What is your favorite subject?  \
14                             Wechat                             CS   
16                           Disney +                           Math   
17                           iMessage                           Math   
24                            Youtube                           Math   
25                             Tiktok                     Chemistry    
..                                ...                            ...   
496                          DuoLingo                           Math   
499                          Snapchat

### 🔬 Test Case Checkpoint 🔬

In [41]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df1_treatment' in vars() ), "You do not have a DataFrame called `df1_treatment`."
assert( 'df1_control' in vars() ), "You do not have a DataFrame called `df1_control`."
assert( len(df[ df.index.isin(df1_treatment.index) & df.index.isin(df1_control.index) ]) == 0), "The `df1_treatment` and `df1_control` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df1_treatment.index) & df.index.isin(df2.index) ]) == 0), "The `df1_treatment` and `df2` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df1_control.index) & df.index.isin(df2.index) ]) == 0), "The `df1_control` and `df2` groups have some overlap!  There must be no overlap."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 3.3:

Create a `df2_treatment` group that contains a random selection of 50% of the rows from `df2`.

In [42]:
df2_treatment = df2.sample(frac = 0.5)
df2_treatment

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
467  Jeonghwan  AYL                    1.0                    3.0   
361     Maciej  AYB                    2.0                    5.0   
49       Yunsu  AYG                    1.0                    3.0   
119   Sabarish  AYD                    2.0                    5.0   
223    Nicole   AYG                    2.0                    3.0   
..         ...  ...                    ...                    ...   
517     Lauren  AYP                    1.0                    3.0   
359   Nityanta  AYB                    2.0                    5.0   
167        Adi  AYB                    2.0                    5.0   
300      Erika  AYH                    2.0                    5.0   
260     Bhuvan  AYS                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
467                                          2-Jan                 7UP   
361                                          2-Jan                 7UP   
49                                           2-Jan         Sierra Mist   
119                                          2-Jan                 7UP   
223                                          2-Jan                 7UP   
..                                             ...                 ...   
517                                          2-Jan                 7UP   
359                                          2-Jan                 7UP   
167                                          2-Jan                 7UP   
300                                            0.5         Sierra Mist   
260                                            0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
467             iOS           Orange          Beef     On/After July 1  ...   
361             iOS            Apple          Beef   On/Before June 30  ...   
49              iOS           Orange          Pork     On/After July 1  ...   
119             iOS            Apple          Pork   On/Before June 30  ...   
223             iOS            Apple          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
517             iOS            Apple          Beef   On/Before June 30  ...   
359             iOS           Orange           NaN   On/Before June 30  ...   
167             iOS            Apple          Pork   On/Before June 30  ...   
300             iOS           Orange          Beef     On/After July 1  ...   
260             iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
467                                Starbucks   
361                        Ice Vanilla Latte   
49                       I dont drink coffee   
119         Java Chip Frappacino - Starbucks   
223                         Strawberry Acai    
..                                       ...   
517                                      tea   
359                               Caffe Bene   
167                                    latte   
300                        Iced tea lemonade   
260                              Nitro Boost   

    What is your favorite mobile app? What is your favorite subject?  \
467                           YouTube                             IS   
361                          Snapchat                           Math   
49                            YouTube                        History   
119                         Instagram                           Math   
223                            TikTok                        Reading   
..                                ...                            ...   
517                         instagram                        english   
359                            TikTok                        Science   
167                          snapchat                           econ   
300                           Discord         

## Part 3.4:

Create a `df2_control` group that contains the other half of the rows from `df2`.

In [43]:
df2_control = df2[~df2.index.isin(df2_treatment.index)]
df2_control

Name   Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
0      CA Ana   AL1                    2.0                    5.0   
3       Lekha   AYH                    2.0                    2.0   
6    Esbeida    AYA                    2.0                    5.0   
9      Advika   AYK                    1.0                    3.0   
12      Aimee   AYM                    1.0                    3.0   
..        ...   ...                    ...                    ...   
510   Sahithi   AYF                    1.0                    3.0   
511     Zaina   AYF                    2.0                    3.0   
512  Veronica   AYO                    1.0                    3.0   
514      Bill   AYA                    1.0                    3.0   
519      Bill   AYC                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
0                                            2-Jan         Sierra Mist   
3                                              0.5                 7UP   
6                                            2-Jan                 7UP   
9                                            2-Jan         Sierra Mist   
12                                           2-Jan                 7UP   
..                                             ...                 ...   
510                                          2-Jan                 7UP   
511                                            0.5         Sierra Mist   
512                                          2-Jan                 7UP   
514                                          2-Jan                 7UP   
519                                          2-Jan         Sierra Mist   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
0               iOS            Apple          Pork   On/Before June 30  ...   
3               iOS           Orange          Pork     On/After July 1  ...   
6               iOS           Orange          Beef     On/After July 1  ...   
9               iOS           Orange          Pork     On/After July 1  ...   
12              iOS            Apple          Beef   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
510             iOS            Apple          Pork   On/Before June 30  ...   
511             iOS           Orange          Beef   On/Before June 30  ...   
512             iOS            Apple          Beef   On/Before June 30  ...   
514             iOS           Orange          Beef   On/Before June 30  ...   
519             iOS            Apple          Beef     On/After July 1  ...   

    What is your favorite coffee shop drink?  \
0                                  Cold brew   
3                        pumpkin spice latte   
6                       Iced vanilla coffee    
9                         iced caramel latte   
12                 Cold brew (Pumpkin cream)   
..                                       ...   
510                              Iced Coffee   
511                        iced matcha latte   
512            Vanilla sweet cream cold brew   
514                                 Starbuck   
519                                      NaN   

    What is your favorite mobile app? What is your favorite subject?  \
0                            Snapchat                           Math   
3                             twitter              english (writing)   
6                           Snapchat                            Math   
9                            snapchat                           math   
12                            twitter                Social science    
..                                ...                            ...   
510                          Snapchat                        History   
511                            tiktok                        science   
512                        Pokemon Go                        History   
514                           Discord        

### 🔬 Test Case Checkpoint 🔬

In [44]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( 'df2_treatment' in vars() ), "You do not have a DataFrame called `df2_treatment`."
assert( 'df2_control' in vars() ), "You do not have a DataFrame called `df2_control`."
assert( len(df[ df.index.isin(df2_treatment.index) & df.index.isin(df2_control.index) ]) == 0), "The `df2_treatment` and `df2_control` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df2_treatment.index) & df.index.isin(df1.index) ]) == 0), "The `df2_treatment` and `df1` groups have some overlap!  There must be no overlap."
assert( len(df[ df.index.isin(df2_control.index) & df.index.isin(df1.index) ]) == 0), "The `df2_control` and `df1` groups have some overlap!  There must be no overlap."


## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Part 3.5:

Finally, read the guide: **Combining DataFrames by Concatenation** (https://discovery.cs.illinois.edu/guides/Combining-DataFrames/Combining-DataFrames-by-Concatenation/) to learn how to combine `df1_treatment` and `df2_treatment` together into one group called `treatment`:

In [45]:
treatment = pd.concat([df1_treatment, df2_treatment])
treatment

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
482    Tucker  AYB                    1.0                    5.0   
102  Shengwen  AYN                    2.0                    5.0   
420    Amanda  AYB                    2.0                    5.0   
207     Jimmy  AYC                    1.0                    5.0   
230   Pickles  AYC                    2.0                    5.0   
..        ...  ...                    ...                    ...   
517    Lauren  AYP                    1.0                    3.0   
359  Nityanta  AYB                    2.0                    5.0   
167       Adi  AYB                    2.0                    5.0   
300     Erika  AYH                    2.0                    5.0   
260    Bhuvan  AYS                    2.0                    3.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
482                                            0.5                 7UP   
102                                          2-Jan                 7UP   
420                                          2-Jan         Sierra Mist   
207                                          2-Jan                 NaN   
230                                            0.5                 7UP   
..                                             ...                 ...   
517                                          2-Jan                 7UP   
359                                          2-Jan                 7UP   
167                                          2-Jan                 7UP   
300                                            0.5         Sierra Mist   
260                                            0.5                 7UP   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
482             iOS           Orange          Beef     On/After July 1  ...   
102             iOS            Apple          Pork   On/Before June 30  ...   
420             iOS           Orange          Beef   On/Before June 30  ...   
207             iOS           Orange          Beef     On/After July 1  ...   
230             iOS            Apple          Pork   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
517             iOS            Apple          Beef   On/Before June 30  ...   
359             iOS           Orange           NaN   On/Before June 30  ...   
167             iOS            Apple          Pork   On/Before June 30  ...   
300             iOS           Orange          Beef     On/After July 1  ...   
260             iOS            Apple          Pork   On/Before June 30  ...   

    What is your favorite coffee shop drink?  \
482                                  frappe    
102                              Cappuccino    
420                                   matcha   
207                                      NaN   
230                                      Tea   
..                                       ...   
517                                      tea   
359                               Caffe Bene   
167                                    latte   
300                        Iced tea lemonade   
260                              Nitro Boost   

    What is your favorite mobile app? What is your favorite subject?  \
482                            TikTok                           Math   
102                            Canvas               Computer science   
420                         smash hit                       business   
207                            WeChat                        Physics   
230                      Illinois app                        English   
..                                ...                            ...   
517                         instagram                        english   
359                            TikTok                        Science   
167                          snapchat                           econ   
300                           Discord                    

## Part 3.6:

Using the same technique, combine `df1_control` and `df2_control` together into one group called `control`:

In [46]:
control = pd.concat([df1_control, df2_control])
control

Name  Lab  Number 1 or Number 2?  Number 3 or Number 5?  \
14   Tianxiang  AYR                    1.0                    5.0   
16     Claudia  AL1                    2.0                    3.0   
17        Wing  AYB                    1.0                    5.0   
24       Brian  AYF                    1.0                    3.0   
25   Shreshika  AYB                    2.0                    3.0   
..         ...  ...                    ...                    ...   
510    Sahithi  AYF                    1.0                    3.0   
511      Zaina  AYF                    2.0                    3.0   
512   Veronica  AYO                    1.0                    3.0   
514       Bill  AYA                    1.0                    3.0   
519       Bill  AYC                    1.0                    5.0   

    What is your preferred way of saying one half? 7UP or Sierra Mist?  \
14                                           2-Jan                 7UP   
16                                           2-Jan         Sierra Mist   
17                                           2-Jan         Sierra Mist   
24                                             0.5                 7UP   
25                                           2-Jan                 7UP   
..                                             ...                 ...   
510                                          2-Jan                 7UP   
511                                            0.5         Sierra Mist   
512                                          2-Jan                 7UP   
514                                          2-Jan                 7UP   
519                                          2-Jan         Sierra Mist   

    Android or iOS? Apple or Orange? Beef or Pork? Your birthday is...  ...  \
14          Android           Orange          Pork     On/After July 1  ...   
16              iOS            Apple          Beef     On/After July 1  ...   
17              iOS            Apple          Beef   On/Before June 30  ...   
24              iOS            Apple          Beef   On/Before June 30  ...   
25              iOS           Orange           NaN   On/Before June 30  ...   
..              ...              ...           ...                 ...  ...   
510             iOS            Apple          Pork   On/Before June 30  ...   
511             iOS           Orange          Beef   On/Before June 30  ...   
512             iOS            Apple          Beef   On/Before June 30  ...   
514             iOS           Orange          Beef   On/Before June 30  ...   
519             iOS            Apple          Beef     On/After July 1  ...   

              What is your favorite coffee shop drink?  \
14                                           Starbucks   
16   Ice caramel macchiato or salted caramel cream ...   
17                                  Bridgeport Coffees   
24                                          Cappuccino   
25                                                 NaN   
..                                                 ...   
510                                        Iced Coffee   
511                                  iced matcha latte   
512                      Vanilla sweet cream cold brew   
514                                           Starbuck   
519                                                NaN   

    What is your favorite mobile app? What is your favorite subject?  \
14                             Wechat                             CS   
16                           Disney +                           Math   
17                           iMessage                           Math   
24                            Youtube                           Math   
25                             Tiktok                     Chemistry    
..                                ...                            ...   
510                          Snapchat                        History   
511                            tiktok                        science  

### 🔬 Test Case Checkpoint 🔬

In [47]:
## == CHECKPOINT TEST CASES ==
# - This read-only cell contains test cases for your previous cell.
# - If this cell runs without any error our output, you PASSED all test cases!
# - If this cell results in any errors, check you previous cell, make changes, and RE-RUN your code and then this cell.
assert( len(df[ df.index.isin(treatment.index) & df.index.isin(control.index) ]) == 0), "The `treatment` and `control` groups have some overlap!  There must be no overlap."

## == SUCCESS MESSAGE ==
# You will only see this message (with the emoji showing) if you passed all test cases:
tada = "\N{PARTY POPPER}"
print(f"{tada} All tests passed! {tada}")

🎉 All tests passed! 🎉


## Analysis: Blocking

**Q: In your own words, explain what blocking is and why it is useful to someone who has never taken a statistics course before! You can use what you did above as an example in your explanation if you'd like. Make sure to include what blocking is and why it is useful. Write at least three complete sentences to answer this question.**

Blocking is done before a randomized control experiment and it is used when you have a relatively small sample, in order to eliminate possible confounders. Blocking isn't utilized for larger samples because in these cases, random differences usually average out.

<hr style="color: #DD3403;">

# Submission

You're almost done!  All you need to do is to commit your lab to GitHub:

1.  ⚠️ **Make certain to save your work.** ⚠️ To do this, go to **File => Save All**

2.  After you have saved, exit this notebook and follow the Canvas instructions to commit this lab to your Git repository!

3. Your TA will grade your submission and provide you feedback after the lab is due. :)